Translation memory for realz

camertron opened the pull request:
Use the Smartling-provided translation memory (via the industry-standard .tmx file format) instead of our hacky one. This is possible because Smartling now includes keys (we call them “meta keys,” Smartling calls them “variants”) in their .tmx dumps.

@lumoslabs/platform

camertron committed:
Download translation memories directly from smartling

camertron committed:
Move smartling and concurrent-ruby to gemspec

camertron committed:
Small improvements to TmxDocument

camertron committed:
Added specs for TmxDocument

jdoconnor commented on Gemfile:
```diff
@@ -2,8 +2,7 @@ source 'https://rubygems.org'
 
-gem 'smartling'
-gem 'concurrent-ruby', '~> 0.7.0'
+gemspec
```
hooray!!

jdoconnor commented on lib/rosette/integrations/smartling-integration/tmx_document.rb:
```diff
@@ -0,0 +1,72 @@
+
+        def start_element(name, attrs = [])
+          case name
+            when 'tu'
```
are these standard element names in a tmx file? If so, consider having them be constants

jdoconnor commented on lib/rosette/integrations/smartling-integration/translation_memory_builder.rb:
```diff
@@ -47,76 +38,42 @@ def set_logger(logger)
-              hash_mutex.synchronize do
-                ret[meta_key] = recent_key
-              end
+              memory_hash = download_and_process(locale)
```
can you call build_synchronously from here?

camertron replied:
Unfortunately not, since `build_synchronously` and `build_asynchronously` both loop through locales. In the case of `build_asynchronously`, a bunch of `Proc`s are added to a thread pool, which `build_synchronously` doesn't do. The `download_and_process` method handles building for a single locale, which both methods call.

jdoconnor replied:
nuts. I was hoping it was an easy case to get down to a single code path

camertron replied:
Yeah me too... for the most part the only repeated code is the call to `#download_and_process`, which is pretty minimal I'd say.

jdoconnor commented on spec/tmx_document_spec.rb:
```diff
@@ -0,0 +1,97 @@
+      }
+
+      parser.parse(tmx_contents)
+      expect(result_hash).to eq({})
```
should this error or return a blank set?

camertron replied:
I was thinking it should return a blank set, since you always download and parse a .tmx file in the context of a locale. This behavior is mostly incidental however because .tmx files have entries for both "en-US" and the target locale (in this case, "de-DE"), and the code tries to ignore the "en-US" locale. It makes sense to me for this method to throw an error though because there should always be an entry for the locale in each `<tu>` element. If not, then something obviously went wrong on the Smartling side. Rather than pull potentially incorrect data, we should probably halt the pulling process.

jdoconnor commented:
looks good. I have a few small questions, but this is great

camertron committed:
Move tmx tag and attribute names into constants

camertron committed:
Raise errors on missing translation units

camertron committed:
Using tmx-parser

camertron committed:
Add SmartlingTmxParser class

Wraps TmxParser, handles smartling-specific properties, etc.

camertron committed:
Remove smartling-specific parsing logic from TranslationMemoryBuilder

Has been moved to SmartlingTmxParser.

camertron committed:
Fix puller so it catches PlaceholderMismatchErrors, etc

Also should pass the full phrase to TranslationMemory#translation_for.

camertron committed:
Fix translation lookup logic and tests

camertron committed:
Require tmx-parser in spec helper

camertron committed:
Fix nil bug in uploader

Directives can be nil.

camertron committed:
Replace inline placeholders

camertron committed:
Remove pry require

camertron committed:
Fixing issue when trying to glue a false onto a string

camertron committed:
Find inline placeholders even when they aren't explicitly delimited in the tmx file

camertron committed:
Don't rescue PlaceholderMismatchError

camertron committed:
Refactor placeholder resolver methods

camertron committed:
Removing another *$#*&*@#% pry-nav require

camertron committed:
Oops forgot to set rosette_config instance variable

camertron committed:
Make resolution check key as well as meta key

This helps disambiguate when there are duplicates of the meta key
(usually this happens because of changes in placeholder identification).

camertron committed:
Report a warning if a translation can't be found

camertron committed:
Allow fuzzy key matching

Fuzzy key matching means the meta key matches exactly but the key
doesn’t have to match. We might consider using like a Levenshtein
distance algorithm for this in the future, but there are going to be a
relatively small number of keys that won’t match, so I think we’re ok
for now.

camertron committed:
Try to fix converting a nil into a string

camertron committed:
Don't report warnings, just log

Too noisy otherwise

camertron committed:
Another stab at fixing the nil into string issue

camertron committed:
Print translation missing messages at warn level instead of info

camertron committed:
Add a naïve cache around translation lookup

camertron committed:
Avoid nils in catch fetch

camertron committed:
Redefine fetch logic

Should not re-compute if translation is nil.

camertron committed:
Remove logging

camertron committed:
Update phrase count in pusher

Somehow this got overlooked, so commit-cron has been mistakenly marking
commits as translated that shouldn’t be.

camertron commented:
Hey @jdoconnor this has changed significantly, would you mind taking another look?

jdoconnor commented on Gemfile:
```diff
@@ -2,10 +2,10 @@ source 'https://rubygems.org'
+gemspec
 
 group :development, :test do
+  gem 'tmx-parser', path: '~/workspace/tmx-parser'
```
just a note that this needs changing prior to merge

camertron replied:
Good catch, I'll publish the tmx-parser gem.

jdoconnor commented on lib/rosette/integrations/smartling-integration/translation_memory.rb:
```diff
@@ -1,43 +1,237 @@
+        def fetch(phrase)
+          key = (phrase.meta_key || '') + (phrase.key || '')
+          translation_cache.fetch(key) do
+            @cache_mutex.synchronize do
```
would it make sense to put a mutex on the key instead the whole cache?

camertron replied:
Yes, how would I do that? Do I need a separate mutex for each key?

jdoconnor replied:
yep. with redis, ensure that the mutex has a ttl (expiry) and they'll happily be marked to expire after a period of time and treat it like a cache key

jdoconnor replied:
if the mutex supports a key do something like "mutex_#{key}"

camertron replied:
Unfortunately rosette uses memcached instead of redis, but even if it didn't, there's no way for rosette-integration-smartling to know which kind of cache is being used. We could write our own cache-agnostic mutex that might look something like [this](https://gist.github.com/torkale/2762063). Thoughts?

jdoconnor replied:
Not needed... Memcache has a TTL as well. I was using that as an example
On Mar 19, 2015 5:22 PM, "Cameron Dutro" notifications@github.com wrote:

> In lib/rosette/integrations/smartling-integration/translation_memory.rb
> https://github.com/rosette-proj/rosette-integration-smartling/pull/4#discussion_r26803880
> :
> 
> > -            translations.keys.sort.each do |meta_key|
> > -              digest << meta_key
> > -              digest << translations[meta_key].key
> > -        private
> >   +
> > -        def fetch(phrase)
> > -          key = (phrase.meta_key || '') + (phrase.key || '')
> > -          translation_cache.fetch(key) do
> > -            @cache_mutex.synchronize do
> 
> Unfortunately rosette uses memcached instead of redis, but even if it
> didn't, there's no way for rosette-integration-smartling to know which kind
> of cache is being used. We could write our own cache-agnostic mutex that
> might look something like this https://gist.github.com/torkale/2762063.
> Thoughts?
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/rosette-proj/rosette-integration-smartling/pull/4/files#r26803880
> .

camertron replied:
Right, but the TTL isn't the issue. We need some cache-agnostic way of locking a single key, which ActiveSupport::Cache doesn't support out-of-the-box.

jdoconnor commented on lib/rosette/integrations/smartling-integration/translation_memory.rb:
```diff
@@ -1,43 +1,237 @@
+        private
+
+        def fetch(phrase)
+          key = (phrase.meta_key || '') + (phrase.key || '')
```
`key = "#{phrase.meta_key}#{phrase.key}"` may read easier

camertron replied:
Lol I are dumb.

jdoconnor commented on lib/rosette/integrations/smartling-integration/translation_memory.rb:
```diff
@@ -1,43 +1,237 @@
+        def find_plural_translation_for(locale, meta_key)
+          all_translations = all_translations_for(locale)
+          meta_key_base = meta_key.sub(PLURAL_REGEX, '')
+          plural_form = $1
```
because this is so far (code wise) from the definition of the regex, consider adding a comment here about what will be returned in the $1 capture

jdoconnor commented on lib/rosette/integrations/smartling-integration/translation_memory_builder.rb:
```diff
@@ -124,91 +82,29 @@ def drain_pool(pool)
-            .set_serializer_id(serializer_id)
-            .set_smartling_api(smartling_api)
+          RestClient.get(uri.to_s).body
+          # File.read("/Users/cameron/Downloads/memories/#{locale.code}-published.tmx")
```
do you want to keep this?

jdoconnor commented on spec/smartling_pusher_spec.rb:
```diff
@@ -66,7 +72,7 @@ def add_file_to_repo
       repo_name: repo_name,
       commit_id: commit_id,
-      phrase_count: 1,
+      phrase_count: nil,
```
is nil the same as zero?

camertron replied:
No, `nil` means the field hasn't been filled out yet (attempting to mimic ActiveRecord here).

jdoconnor commented:
looking awesome. I had a few small questions, but consider this a +1 when they're considered.

camertron committed:
Making changes suggested by @jdoconnor
